tests/end_to_end/test_import_submodule.py

```diff
@@ -0,0 +1,106 @@
+#     os.chdir(tmp_path)
+
+
+def test_error_not_imported(tmp_path, execute):
+    """
+    Verify that trying to access a not imported submodule raises an AttributeError
+    """
+    shutil.copytree("tests/end_to_end/import_data", tmp_path / "import_data")
+    os.chdir(tmp_path)
+
+    code = """import import_data.utils
+is_prime = import_data.utils.__no_imported_submodule.is_prime
+"""
+    with pytest.raises(UserException):
+        execute(code, snapshot=False)
+
+
+def test_from(tmp_path, execute):
+    """
+    Test import a submodule from a parent
+    """
+    shutil.copytree("tests/end_to_end/import_data", tmp_path / "import_data")
+    os.chdir(tmp_path)
+    code = """from import_data.utils import __no_imported_submodule
+is_prime = __no_imported_submodule.is_prime
+"""
+    res = execute(code, artifacts=["is_prime"])
+    assert res.values["is_prime"] is False
+    assert res.artifacts["is_prime"] == code
+
+
+def test_direct(tmp_path, execute):
+    """
+    Test importing a submodule and referring to it directly
+    """
+    shutil.copytree("tests/end_to_end/import_data", tmp_path / "import_data")
+    os.chdir(tmp_path)
+
+    code = """import import_data.utils.__no_imported_submodule
+is_prime = import_data.utils.__no_imported_submodule.is_prime
+"""
+    res = execute(code, artifacts=["is_prime"])
+    assert res.values["is_prime"] is False
+    assert res.artifacts["is_prime"] == code
+
+
+def test_slice_parent(tmp_path, execute):
+    """
+    Test that slicing on a submodule should remove the import for the parent
+    """
+    shutil.copytree("tests/end_to_end/import_data", tmp_path / "import_data")
+    os.chdir(tmp_path)
+
+    code = """import import_data.utils.__no_imported_submodule
+import import_data.utils
+is_prime = import_data.utils.__no_imported_submodule.is_prime
+"""
+    res = execute(code, artifacts=["is_prime"])
+    assert res.values["is_prime"] is False
+    assert (
+        res.artifacts["is_prime"]
+        == """import import_data.utils.__no_imported_submodule
+is_prime = import_data.utils.__no_imported_submodule.is_prime
+"""
+    )
+
+
+def test_slice_sibling(tmp_path, execute):
+    """
+    Test importing a submodule and referring to it directly, should have a different slice
+    """
+    shutil.copytree("tests/end_to_end/import_data", tmp_path / "import_data")
+    os.chdir(tmp_path)
+
+    code = """import import_data.utils.__no_imported_submodule
+import import_data.utils.__no_imported_submodule_prime
+first_is_prime = import_data.utils.__no_imported_submodule.is_prime
+second_is_prime = import_data.utils.__no_imported_submodule_prime.is_prime
+"""
+    res = execute(code, artifacts=["first_is_prime", "second_is_prime"])
+    assert res.values["first_is_prime"] is False
+    assert res.values["second_is_prime"] is True
+    assert (
+        res.artifacts["first_is_prime"]
+        == """import import_data.utils.__no_imported_submodule
+import import_data.utils.__no_imported_submodule_prime
```

Comment:
I don't understand why this line is needed for this particular use case. I verified that the value for `first_is_prime` does not change based on whether `import import_data.utils.__no_imported_submodule_prime` is included.

<img width="619" alt="Screen Shot 2022-05-12 at 11 08 39 PM" src="https://user-images.githubusercontent.com/504219/168221621-9a1130bd-e53e-415f-afaa-7dd853089ef0.png">

The only reason would be that we are pessimistically assuming some weird mutation. In which case, I'd like to at least know what real world libraries has that behavior for us to warrant using such a heuristic. I think @lionsardesai mentioned something about Matplotlib? Could we add a test for that/make our own test follow the pattern?

Reply:
discussed with yifan, we are pessimistically assuming that all imports mutate the parent package. so whether or not there is mutation, we are saying the __no_imported_submodule mutated the import_data.utils package. so when you want to access something in __no_imported_submodule_prime, we still import the first one because it _could_ have imported utils and we want to build on that.